### ISSUE TITLE
From tf frozen graph to onnx tf2onnx.convert failed to export padding size in correct order

### ISSUE THREAD

**Issue by stonexjr** [onnx/tensorflow-onnx#201]
OK, I tried to make it short. 
To recreate the issue, do as follows:
1. Download the public pre-trained Resnet-50 v2 model from tensorflow git [repo](https://github.com/tensorflow/models/tree/master/official/resnet), select the one with ResNet-50 v2, fp32 , savedModel [NHWC](http://download.tensorflow.org/models/official/20181001_resnet/savedmodels/resnet_v2_fp32_savedmodel_NHWC.tar.gz)
2. Use `tensorflow.python.tools.freeze_graph` to export a frozen graph from savedModel (see instruction [here](https://github.com/tensorflow/models/tree/master/research/slim#freezing-the-exported-graph)).  Here is the command I used to create frozen graph from savedModel
```bash
bazel-bin/tensorflow/python/tools/freeze_graph \
--input_saved_model_dir="/root/models/resnet50/tensorflow/resnet_v2_fp32_savedmodel_NHWC/1538687283" \ 
--output_graph="/root/models/resnet50/tensorflow/frozen_graph/resnet50_v2_frozen_graph.pb" \
--output_node_names="softmax_tensor"
```
For your convenience here is my pregenerated [frozen graph](https://drive.google.com/open?id=1Yqml6YIoNe5g55isfkSBgIqR4exexxmP)

3. Now use tf2onnx.convert to convert the frozen graph from step (2) to onnx format. 
Here is the command I used to create onnx from frozen graph
```bash
INPUT_FROZEN_GRAPH_PB="/root/models/resnet50/tensorflow/frozen_graph/resnet50_v2_frozen_graph.pb"
OUTPUT_ONNX_FILE="/root/models/resnet50/onnx/resnet50_v2.onnx"
INPUTS="input_tensor:0"
OUTPUTS="softmax_tensor:0"
# --inputs-as-nchw $INPUTS # It doesn't make any difference with this line regarding this issue.
python3 -m tf2onnx.convert \
 --input  $INPUT_FROZEN_GRAPH_PB \
 --inputs $INPUTS \
 --outputs $OUTPUTS \
 --output $OUTPUT_ONNX_FILE \
 --opset 7 \
 --verbose
```
For your convenience here is my pregenerated [onnx](https://drive.google.com/open?id=1GGi1E7Kz8Tw3subge3xeWYxxW-J9SxW6).

4. Use `onnx2trt` from [onnx-tensort](https://github.com/onnx/onnx-tensorrt) to load and parse the onnx model from (3). Then you will see the error message: 
```
Parsing ONNX file...
----------------------------------------------------------------
Input filename:  /root/models/resnet50/onnx/resnet50_v2.onnx
ONNX IR version:  0.0.3
Opset version:    7
Producer name:    tf2onnx
Producer version: 0.4.0
Domain:
Model version:    0
Doc string:
----------------------------------------------------------------
While parsing node number 5 [Pad -> "resnet_model/Pad:0"]:
ERROR: /home/erisuser/p4sw/sw/gpgpu/MachineLearning/DIT/release/5.0/parsers/onnxOpenSource/builtin_op_importers.cpp:1135 In function importPad:
[8] Assertion failed: onnx_padding[0] == 0 && onnx_padding[1] == 0 && onnx_padding[4] == 0 && onnx_padding[5] == 0
ERROR: failed to parse onnx file
```

Thoughts
I visualized graph topologies from both frozen graph and its onnx couterparts. Apparently, the padding informations (i.e. padding 3 pixels at left, right, top and bottom of input image) get screwed up in step (3) 

Here is what the padding node looks like in the original frozen graph

<img width="712" alt="2 padding size in original frozen graph" src="https://user-images.githubusercontent.com/2969316/48808465-bc87b680-ecd5-11e8-86aa-5e211d3559c4.png">

And here is what it looks like after conversion in step (3)
<img width="522" alt="1 incorrect padding size after conversion" src="https://user-images.githubusercontent.com/2969316/48808471-c3aec480-ecd5-11e8-885f-a5b92cff11dc.png">

**Comment by guschmue** [comment]
looking ...

**Comment by guschmue** [comment]
After looking at it ... I think tensorflow-onnx does the correct thing. A bit lengthy but here you go:

TF says:
https://www.tensorflow.org/api_docs/python/tf/pad
For each dimension D of input, paddings[D, 0] indicates how many values to add before the contents 
of tensor in that dimension, and paddings[D, 1] indicates how many values to add after the contents 
of tensor in that dimension.

Onnx says:
https://github.com/onnx/onnx/blob/master/docs/Operators.md#Pad
[x1_begin, x2_begin...x1_end, x2_end,...], where xi_begin the number of pixels added
at the beginning of axis `i` and xi_end, the number of pixels added at the end of axis `i`

If TF gives us the padding:
```[[0, 0],  [3, 3], [3, 3], [0, 0]]```

It should map to onnx like:
```[0, 3, 3, 0, 0, 3, 3, 0]```
(basically take all even and concat all odd list indices)

That is what tensorflow-onnx fills in.

If I validate this by looking at the shapes that the tensorflow graph has:
```IN: [64,224,224,3] -> OUT: [64,230,230,3]```

```
for tensorflow-onnx:
[0, 3, 3, 0, 0, 3, 3, 0]
[0+64+0,3+224+3,3+224+3,0+3+0] -> correct

for tensorrt:
[0, 0, 3, 3, 0, 0, 3, 3]
[0+64+0,0+224+0,3+224+3,3+3+3] -> not correct
```
We could try to find a tensorrt person to help with this.

**Comment by pengwa** [comment]
if you feel strentchy looking into transpose optimizer, I can help, but probably have to put it as lower priority now, maybe next week. :)

**Comment by stonexjr** [comment]
Thank you all for looking into this. I've also ping folks in the onnx-tensorflow repo [here](https://github.com/onnx/onnx-tensorrt/issues/79).

**Comment by guschmue** [comment]
no worry, pretty sure we can take of it.
@pengwa - thanks!! I have some time today, hope its a trivial thing.

**Comment by guschmue** [comment]
@pengwa I have a simple unit test and see the problem - thinking about a generic fix, something like fusing the transpose with the previous op.

**Comment by guschmue** [comment]
have not forgotten - planning to fuse the pads into conv() ... just need to find time.

**Comment by guschmue** [comment]
Yes, its not a bug in tensorflow-onnx.
Moving the tf.pad into conv() would resolve it and would give us some perf gain ... I think there is even an open issue asking for it.

**Comment by pengwa** [comment]
got it! sound great! :)

**Comment by pengwa** [comment]
the pad + conv fusion is done by @lucienwang1009, let us know if you still have issue on tf2onnx side.
